src/ClientIp.php

```diff
@@ -107,33 +107,70 @@ public function process(ServerRequestInterface $request, DelegateInterface $dele
      */
     private function getIp(ServerRequestInterface $request)
     {
-        if ($this->remote) {
-            $ip = file_get_contents('http://ipecho.net/plain');
+        if ($ip = $this->getRemoteIp()) {
+            return $ip;
+        }
 
-            if (self::isValid($ip)) {
+        if ($ip = $this->getProxyIp($request)) {
+            if (!in_array($ip, $this->proxyIps)) {
                 return $ip;
             }
         }
 
-        $ip = null;
-        $server = $request->getServerParams();
-
-        if (!empty($server['REMOTE_ADDR']) && self::isValid($server['REMOTE_ADDR'])) {
-            $ip = $server['REMOTE_ADDR'];
+        if ($ip = $this->getLocalIp($request)) {
+            if (!in_array($ip, $this->proxyIps)) {
+                return $ip;
+            }
```

Comment:
You may want to configure the proxies with or without ips. No ips means you're trusting in all proxies, so just try to get the ip from headers and use the remote_addr as fallback. Obviously, this less secure. Setting an array of ips of well known proxies, get the ip only from any of these proxies, and use remote_addr as fallback too.